### PR TITLE
Fix misc. loader and player issues found by libFuzzer.

### LINKED ISSUE #400: Usage of integer division.
Several places in libxmp use integer division, which is slow and very unfriendly to platforms without a dedicated integer division instruction like ARMv6 (and are still slow even if there is one). Unlike libxmp's floating point usage, this seems to have creeped into several inner loops. Note this only applies to non-literal divisors (literal divisors get optimized to a MUL+SHR).

Roughly in decreasing order of urgency:

* `mixer.c` uses an integer division once per sample frame for declicking(!). This only affects small portions of samples but should be fixed. (#404)
* The following loaders use integer division for pattern events instead of two for loops for some reason:
  * `mod_load.c` (#457)
  * `st_load.c` (#609)
  * `xm_load.c` (#461)
  * ~~`abk_load.c`~~ effect conversion
  * `asylum_load.c` (#501)
  * `coco_load.c` (#443)
  * `far_load.c` (#500)
  * ~~`flt_load.c`~~ AM instrument calculation
  * `fnk_load.c` (#446)
  * `okt_load.c`
  * `sfx_load.c`
  * `stm_load.c` (#482)
* `player.c` uses these when computing envelopes (probably reworkable) and when computing the current tick (can be manually unswitched so it's guaranteed to only be performed once per tick).
* `mix_paula.c`
* Misc. depackers.

I found these by searching a disassembly of libxmp.dll for any places `idiv*` was emitted, which may not be foolproof.

### PR DESCRIPTION
**Pull request by AliceLR**
This patch fixes some bugs I found fuzzing other recent patches, as well as several existing player issues I found in my corpus after adding basic playback testing to the libFuzzer frontend and my [test module loader](https://gist.github.com/AliceLR/58fadf18f2248e163faed113baa0a25d). (To reduce the noise, I'm experimenting with collapsing HTML in this PR to hide the annoying fuzzer logs. Click them to expand.)

* Unrelated fix: while trying to remember what exactly "psm_load.c" is (it's old format Epic MegaGames MASI...), I found a minor bug in the Protracker 3.6 loader that prevented loading valid modules.
* Unrelated fix 2: I fixed the part of #400 for the ASYLUM loader here while I was working on that loader. It always has 8 channels, so it doesn't need an extra pattern loop.

I haven't done any actual fuzzing with the updated fuzzers yet, so expect more where these came from (:

### Player fixes

<details><summary>Out-of-bounds read caused by missing check on subinstrument sample IDs in `read_event_ft2` (segmentation fault).</summary>

```
Loading 'test-dev/CORPUS/fbd1ef9c441fbb56b74671de908694eeaee94747'... mem:0 file(test):0=================================================================
==185945==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x616000028ad0 at pc 0x7f7813b17404 bp 0x7ffed1c3ab90 sp 0x7ffed1c3ab88
READ of size 4 at 0x616000028ad0 thread T0
    #0 0x7f7813b17403 in read_event_ft2 /home/a/libxmp/src/read_event.c:674:22
    #1 0x7f7813b13ee4 in libxmp_read_event /home/a/libxmp/src/read_event.c:1626:10
    #2 0x7f7813b0c440 in read_row /home/a/libxmp/src/player.c:536:5
    #3 0x7f7813b0abf5 in xmp_play_frame /home/a/libxmp/src/player.c:1722:3
    #4 0x4fde37 in main /home/a/libxmp/mem.cpp:135:12
    #5 0x7f7813591b74 in __libc_start_main (/lib64/libc.so.6+0x27b74)
    #6 0x41d51d in _start (/home/a/libxmp/mem+0x41d51d)

0x616000028ad0 is located 32 bytes to the right of 560-byte region [0x616000028880,0x616000028ab0)
allocated by thread T0 here:
    #0 0x4c3657 in calloc (/home/a/libxmp/mem+0x4c3657)
    #1 0x7f7813b6b1cc in get_chunk_is /home/a/libxmp/src/loaders/mdl_load.c:748:38
    #2 0x7f7813b48c51 in iff_process /home/a/libxmp/src/loaders/iff.c:52:8
    #3 0x7f7813b487a1 in iff_chunk /home/a/libxmp/src/loaders/iff.c:124:9
    #4 0x7f7813b484bc in libxmp_iff_load /home/a/libxmp/src/loaders/iff.c:148:9
    #5 0x7f7813b674e7 in mdl_load /home/a/libxmp/src/loaders/mdl_load.c:1195:9
    #6 0x7f7813b3dbab in load_module /home/a/libxmp/src/load.c:260:18
    #7 0x7f7813b3e19e in xmp_load_module_from_memory /home/a/libxmp/src/load.c:442:8
    #8 0x4fdbfd in main /home/a/libxmp/mem.cpp:88:14
    #9 0x7f7813591b74 in __libc_start_main (/lib64/libc.so.6+0x27b74)

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/a/libxmp/src/read_event.c:674:22 in read_event_ft2
Shadow bytes around the buggy address:
  0x0c2c7fffd100: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2c7fffd110: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c2c7fffd120: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c2c7fffd130: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c2c7fffd140: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0c2c7fffd150: 00 00 00 00 00 00 fa fa fa fa[fa]fa fa fa fa fa
  0x0c2c7fffd160: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2c7fffd170: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2c7fffd180: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2c7fffd190: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2c7fffd1a0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==185945==ABORTING
```
```
Loading 'test-dev/CORPUS/ff8a5a7457972a84f5b98c28508bcaaf81c54851'... mem:0 file(test):0MemorySanitizer:DEADLYSIGNAL
==139399==ERROR: MemorySanitizer: SEGV on unknown address 0x000000000058 (pc 0x7fa3ddbf6b6e bp 0x7fffea77ca90 sp 0x7fffea77c960 T139399)
==139399==The signal is caused by a READ memory access.
==139399==Hint: address points to the zero page.
    #0 0x7fa3ddbf6b6e in read_event_ft2 /home/a/libxmp/src/read_event.c:674:22
    #1 0x7fa3ddbf2bfa in libxmp_read_event /home/a/libxmp/src/read_event.c:1626:10
    #2 0x7fa3ddbe80d2 in read_row /home/a/libxmp/src/player.c:536:5
    #3 0x7fa3ddbe6638 in xmp_play_frame /home/a/libxmp/src/player.c:1722:3
    #4 0x4a16f1 in main /home/a/libxmp/mem.cpp:135:12
    #5 0x7fa3dd67db74 in __libc_start_main (/lib64/libc.so.6+0x27b74)
    #6 0x41e49d in _start (/home/a/libxmp/mem+0x41e49d)

MemorySanitizer can not provide additional info.
SUMMARY: MemorySanitizer: SEGV /home/a/libxmp/src/read_event.c:674:22 in read_event_ft2
==139399==ABORTING
```
</details>

<details><summary>Heap overflow when playing 0 channel modules that contain an invalid pattern at order 0 (segmentation fault).</summary>

```
Loading 'test-dev/CORPUS/f0cc305d2a6aba7474f6c98e95c6455bd87b3ebf'... mem:0 file(test):0src/mixer.c:330:16: runtime error: inf is outside the range of representable values of type 'int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior src/mixer.c:330:16 in 
src/mixer.c:586:8: runtime error: signed integer overflow: -2147483648 * 2 cannot be represented in type 'int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior src/mixer.c:586:8 in 
 play:0
```
```
Loading 'test-dev/CORPUS/ec7baa27cba13cbd403975381f54a02dec17c75e'... mem:0 file(test):0src/mixer.c:330:16: runtime error: inf is outside the range of representable values of type 'int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior src/mixer.c:330:16 in 
src/mixer.c:586:8: runtime error: signed integer overflow: -2147483648 * 2 cannot be represented in type 'int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior src/mixer.c:586:8 in 
 play:0
```
```
Loading 'test-dev/CORPUS/3f4ace68450432be453ddf98559cb4d140d99210'... mem:0 file(test):0MemorySanitizer:DEADLYSIGNAL
==122090==ERROR: MemorySanitizer: SEGV on unknown address 0x733000020000 (pc 0x7ff44768468d bp 0x7ffc870e2cf0 sp 0x7ffc870e2c70 T122090)
==122090==The signal is caused by a READ memory access.
    #0 0x7ff44768468d in downmix_int_16bit /home/a/libxmp/src/mixer.c
    #1 0x7ff447683284 in libxmp_mixer_softmixer /home/a/libxmp/src/mixer.c:597:3
    #2 0x7ff44764b8a5 in xmp_play_frame /home/a/libxmp/src/player.c:1748:2
    #3 0x4a16f1 in main /home/a/libxmp/mem.cpp:135:12
    #4 0x7ff4470e2b74 in __libc_start_main (/lib64/libc.so.6+0x27b74)
    #5 0x41e49d in _start (/home/a/libxmp/mem+0x41e49d)

MemorySanitizer can not provide additional info.
SUMMARY: MemorySanitizer: SEGV /home/a/libxmp/src/mixer.c in downmix_int_16bit
==122090==ABORTING
```
```
Loading 'test-dev/CORPUS/c03d61a62ac8a12dd0b946347bc05601ffac6bc0'... mem:0 file(test):0=================================================================
==175273==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x633000018824 at pc 0x7fda0630116e bp 0x7fff6d0668b0 sp 0x7fff6d0668a8
READ of size 4 at 0x633000018824 thread T0
    #0 0x7fda0630116d in downmix_int_16bit /home/a/libxmp/src/mixer.c:188:9
    #1 0x7fda06300440 in libxmp_mixer_softmixer /home/a/libxmp/src/mixer.c:597:3
    #2 0x7fda062d9d9e in xmp_play_frame /home/a/libxmp/src/player.c:1748:2
    #3 0x4fde37 in main /home/a/libxmp/mem.cpp:135:12
    #4 0x7fda05d60b74 in __libc_start_main (/lib64/libc.so.6+0x27b74)
    #5 0x41d51d in _start (/home/a/libxmp/mem+0x41d51d)

0x633000018824 is located 0 bytes to the right of 98340-byte region [0x633000000800,0x633000018824)
allocated by thread T0 here:
    #0 0x4c3657 in calloc (/home/a/libxmp/mem+0x4c3657)
    #1 0x7fda06301f55 in libxmp_mixer_on /home/a/libxmp/src/mixer.c:821:23
    #2 0x7fda062d75d6 in xmp_start_player /home/a/libxmp/src/player.c:1524:6
    #3 0x4fddf6 in main /home/a/libxmp/mem.cpp:129:4
    #4 0x7fda05d60b74 in __libc_start_main (/lib64/libc.so.6+0x27b74)

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/a/libxmp/src/mixer.c:188:9 in downmix_int_16bit
Shadow bytes around the buggy address:
  0x0c667fffb0b0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c667fffb0c0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c667fffb0d0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c667fffb0e0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c667fffb0f0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0c667fffb100: 00 00 00 00[04]fa fa fa fa fa fa fa fa fa fa fa
  0x0c667fffb110: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c667fffb120: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c667fffb130: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c667fffb140: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c667fffb150: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==175273==ABORTING
```
</details>

<details><summary>Out-of-bounds read in His Master's Noise Mega-Arp effect (UBSan).</summary>

```
Loading 'test-dev/CORPUS/e7d3f3549224eaea53eefbd3e8acd165278afdec'... mem:0 file(test):0src/hmn_extras.c:134:28: runtime error: index 98 out of bounds for type 'uint8 [16][16]'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior src/hmn_extras.c:134:28 in 
 play:0
```
</details>

<details><summary>Out-of-bounds read in `do_toneporta` caused by missing invalid note checks (UBSan).</summary>

```
Loading 'test-dev/CORPUS/deefa776e249bae5ae38a3b7b2bc270de36e4b92'... mem:0 file(test):0src/effects.c:75:15: runtime error: index -1 out of bounds for type 'struct (anonymous struct at include/xmp.h:222:2) [121]'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior src/effects.c:75:15 in 
 play:0
```
```
Loading 'test-dev/CORPUS/694cf516219d50c511bd4d54284e82d82f202d25'... mem:0 file(test):0src/effects.c:75:15: runtime error: index -1 out of bounds for type 'struct (anonymous struct at include/xmp.h:222:2) [121]'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior src/effects.c:75:15 in 
src/effects.c:87:4: runtime error: index -1 out of bounds for type 'struct (anonymous struct at include/xmp.h:222:2) [121]'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior src/effects.c:87:4 in 
 play:0
```
</details>

<details><summary>Invalid double to integer conversion with low PERIOD_CSPD frequencies (UBSan).</summary>

```
Loading 'test-dev/CORPUS/c3f597024392a8fdfd1b26727514a79d24151197'... mem:0 file(test):0src/player.c:951:20: runtime error: 1.46611e+11 is outside the range of representable values of type 'int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior src/player.c:951:20 in 
 play:0
```
</details>

<details><summary>Invalid double to integer conversion with bad voice positions (GDM) (UBSan).</summary>

Several other cases that could cause this were fixed by adding sample bounding in their respective sample loaders, but GDM sample sizes aren't very well bounded. Bounding bad loops in the sample loader broke numerous loader tests and the fix was very simple, anyway (don't cast to int until after the step size in samples is clamped).

```
Loading 'test-dev/CORPUS/95ccebb1ff35f9ad80b0cd0781f1e1ff8269a969'... mem:0 file(test):0src/mixer.c:479:13: runtime error: 2.20725e+09 is outside the range of representable values of type 'int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior src/mixer.c:479:13 in 
src/mixer.c:556:9: runtime error: signed integer overflow: 80 - -2147483648 cannot be represented in type 'int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior src/mixer.c:556:9 in 
src/player.c:1346:22: runtime error: -2.5565e+09 is outside the range of representable values of type 'int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior src/player.c:1346:22 in 
src/mixer.c:407:14: runtime error: -2.5565e+09 is outside the range of representable values of type 'int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior src/mixer.c:407:14 in 
 play:0
```
</details>

### Loader fixes:

<details><summary>UMR caused by missing return value check on `hio_read` in MASI TITL chunk handler (MSan).</summary>

```
Uninitialized bytes in __interceptor_strlen at offset 4 inside [0x726000072548, 7)
==107681==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x4f0448 in libxmp_adjust_string /home/a/libxmp/test-dev/.fuzzer/src/load_helpers.c:129:18
    #1 0x4f6c29 in load_module /home/a/libxmp/test-dev/.fuzzer/src/load.c:305:2
    #2 0x4f767d in xmp_load_module_from_memory /home/a/libxmp/test-dev/.fuzzer/src/load.c:442:8
    #3 0x4ea779 in LLVMFuzzerTestOneInput /home/a/libxmp/test-dev/libxmp_fuzz.c:15:6
    #4 0x44a739 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/home/a/libxmp/test-dev/libxmp_fuzz_msan+0x44a739)
    #5 0x44b2a0 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) (/home/a/libxmp/test-dev/libxmp_fuzz_msan+0x44b2a0)
    #6 0x44c3b3 in fuzzer::Fuzzer::MutateAndTestOne() (/home/a/libxmp/test-dev/libxmp_fuzz_msan+0x44c3b3)
    #7 0x44dea7 in fuzzer::Fuzzer::Loop(std::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) (/home/a/libxmp/test-dev/libxmp_fuzz_msan+0x44dea7)
    #8 0x43331a in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/home/a/libxmp/test-dev/libxmp_fuzz_msan+0x43331a)
    #9 0x422b46 in main (/home/a/libxmp/test-dev/libxmp_fuzz_msan+0x422b46)
    #10 0x7f3e908a2b74 in __libc_start_main (/lib64/libc.so.6+0x27b74)
    #11 0x422b7d in _start (/home/a/libxmp/test-dev/libxmp_fuzz_msan+0x422b7d)

  Uninitialized value was stored to memory at
    #0 0x49eee2 in strncpy (/home/a/libxmp/test-dev/libxmp_fuzz_msan+0x49eee2)
    #1 0x762138 in get_titl /home/a/libxmp/test-dev/.fuzzer/src/loaders/masi_load.c:135:2
    #2 0x804745 in iff_process /home/a/libxmp/test-dev/.fuzzer/src/loaders/iff.c:52:8
    #3 0x804745 in iff_chunk /home/a/libxmp/test-dev/.fuzzer/src/loaders/iff.c:124:9
    #4 0x804745 in libxmp_iff_load /home/a/libxmp/test-dev/.fuzzer/src/loaders/iff.c:148:9
    #5 0x76133e in masi_load /home/a/libxmp/test-dev/.fuzzer/src/loaders/masi_load.c:754:6
    #6 0x4f646e in load_module /home/a/libxmp/test-dev/.fuzzer/src/load.c:260:18
    #7 0x4f767d in xmp_load_module_from_memory /home/a/libxmp/test-dev/.fuzzer/src/load.c:442:8
    #8 0x4ea779 in LLVMFuzzerTestOneInput /home/a/libxmp/test-dev/libxmp_fuzz.c:15:6
    #9 0x44a739 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/home/a/libxmp/test-dev/libxmp_fuzz_msan+0x44a739)
    #10 0x44b2a0 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) (/home/a/libxmp/test-dev/libxmp_fuzz_msan+0x44b2a0)
    #11 0x44c3b3 in fuzzer::Fuzzer::MutateAndTestOne() (/home/a/libxmp/test-dev/libxmp_fuzz_msan+0x44c3b3)
    #12 0x44dea7 in fuzzer::Fuzzer::Loop(std::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) (/home/a/libxmp/test-dev/libxmp_fuzz_msan+0x44dea7)
    #13 0x43331a in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/home/a/libxmp/test-dev/libxmp_fuzz_msan+0x43331a)
    #14 0x422b46 in main (/home/a/libxmp/test-dev/libxmp_fuzz_msan+0x422b46)
    #15 0x7f3e908a2b74 in __libc_start_main (/lib64/libc.so.6+0x27b74)

  Uninitialized value was created by an allocation of 'buf' in the stack frame of function 'get_titl'
    #0 0x761fb0 in get_titl /home/a/libxmp/test-dev/.fuzzer/src/loaders/masi_load.c:129

SUMMARY: MemorySanitizer: use-of-uninitialized-value /home/a/libxmp/test-dev/.fuzzer/src/load_helpers.c:129:18 in libxmp_adjust_string
Exiting
MS: 1 ShuffleBytes-; base unit: 9b168735916a9c2462ffd9943da9b0c974b44e7b
0x50,0x53,0x4d,0x20,0x5d,0x5,0x1,0x0,0x46,0x49,0x4c,0x45,0x54,0x49,0x54,0x4c,0x45,0xa,0x4c,0x0,0x54,0x49,0x54,0x4c,
PSM ]\x05\x01\x00FILETITLE\x0aL\x00TITL
artifact_prefix='ARTIFACTS/'; Test unit written to ARTIFACTS/crash-9c43703df954a93cc0729c9a81add802c92d7351
Base64: UFNNIF0FAQBGSUxFVElUTEUKTABUSVRM
```
</details>

<details><summary>Signed integer overflow when scanning MED4 samples (UBSan).</summary>

```
src/loaders/med4_load.c:544:10: runtime error: signed integer overflow: -2147483648 - 22 cannot be represented in type 'int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior src/loaders/med4_load.c:544:10 in 
MS: 2 ChangeASCIIInt-CopyPart-; base unit: d1f577df93d7c1372f1913d3326fc15bdf9db324
artifact_prefix='ARTIFACTS/'; Test unit written to ARTIFACTS/crash-d604fe9d22315373cd60723f3ace1061af3f5a9e
```
</details>

<details><summary>Invalid double to integer conversion with bad voice positions (ASYLUM) (UBSan).</summary>

```
Loading 'test-dev/CORPUS/637963bc32d24761aff01f5ef41256a341a6c885'... mem:0 file(test):0src/mixer.c:479:13: runtime error: 5.27511e+09 is outside the range of representable values of type 'int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior src/mixer.c:479:13 in 
```
</details>

<details><summary>Invalid double to integer conversion with bad voice positions (Protracker Studio 16 PSM) (UBSan).</summary>

```
Loading '0c1ae994be07ea4b6df122956be2a1b938d772c5'... mem:0 file(test):0src/player.c:1346:22: runtime error: -1.62825e+10 is outside the range of representable values of type 'int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior src/player.c:1346:22 in
```
</details>

<details><summary>Invalid double to integer conversion with bad voice positions (Ultra Tracker) (UBSan).</summary>

```
Loading 'test-dev/CORPUS/5e997352a2434be81527a0e19accc797a4a925fc'... mem:0 file(test):0src/mixer.c:479:13: runtime error: 2.44609e+09 is outside the range of representable values of type 'int'
```
</details>

### CODE REVIEW

**Comment by AliceLR** [comment]
Marked as a draft so I can figure out what caused the coverage drop...

**Comment by AliceLR** [comment]
I'm getting rate limited by Codecov again, but it *looks* like the coverage drop is fixed now (the new checks broke coverage for later checks that fuzzer files were supposed to fail...).

**Comment by AliceLR** [comment]
There's a bug I missed that ASan found where `f->loop_chn` would persist between modules when reusing the same context. There *might* be another `read_event_ft` issue too, but it can wait until the next patch if this gets merged.

File: src/player.c [hunk 1]
```diff
@@ -1554,8 +1554,11 @@ int xmp_start_player(xmp_context opaque, int rate, int format)
 		mod->len = 0;
 	}
 
-	if (mod->len == 0 || mod->chn == 0) {
+	if (mod->len == 0) {
```
Comment (sezero) [inline src/player.c:1557]: how do this manage leading to crashes?

File: src/hmn_extras.c [hunk 2]
```diff
@@ -131,6 +131,9 @@ void libxmp_hmn_extras_process_fx(struct context_data *ctx, struct channel_data
 {
 	switch (fxt) {
 	case FX_MEGAARP:
+		/* Not sure if this is correct... */
```
Comment (sezero) [inline src/hmn_extras.c:134]: is there a way we can verify this?

File: src/loaders/asylum_load.c [hunk 3]
```diff
@@ -124,26 +137,30 @@ static int asylum_load(struct module_data *m, HIO_HANDLE *f, const int start)
...
 			return -1;
 
-		for (j = 0; j < 64 * mod->chn; j++) {
+		if (hio_read(buf, 1, 2048, f) < 2048) {
+			D_(D_CRIT "read error at pattern %d", i);
+			return -1;
+		}
+
+		pos = buf;
+		for (j = 0; j < 64 * 8; j++) {
```
Comment (sezero) [inline src/loaders/asylum_load.c:150]: is `8` guaranteed here?

Comment (sezero) [reply src/loaders/asylum_load.c:150]: in any case, we may need similar changes in libmikmod.

**Comment by sezero** [comment]
> There _might_ be another `read_event_ft2` issue too, but it can wait until the next patch if this gets merged.

we can wait until you nail that one

File: src/player.c (hunk 1 above)
Comment (AliceLR) [reply src/player.c:1557]: The two blocks of code above this skip invalid patterns at the beginning of the module. These invalid patterns can't be scanned, so their `xxo_info` entries are default values, which includes a BPM value of 0. The function `update_from_ord_info` (called at line 1571) sets the player BPM to the `xxo_info` BPM. The 0 BPM is used to compute `s->ticksize` in mixer.c:330, which resolves to `inf` due to float division by 0. At 584 this is converted to `int`, which seems to specifically result in `INT_MIN`, and then `downmix_int_16bit` tries to iterate on that and reads/writes well past the input/output buffer bounds.

File: src/hmn_extras.c (hunk 2 above)
Comment (AliceLR) [reply src/hmn_extras.c:134]: I don't have the speical version of Noisetracker used for this format *or* an emulation setup that can run older Amiga trackers right now. Claudio might know, but considering that there are only 16 arpeggio patterns, I'm not sure there is a correct handling for the higher values.

File: src/loaders/asylum_load.c (hunk 3 above)
Comment (AliceLR) [reply src/loaders/asylum_load.c:150]: Yes, this format has a fixed 8 channels.

File: src/player.c (hunk 1 above)
Comment (sezero) [reply src/player.c:1557]: ok.

File: src/loaders/asylum_load.c (hunk 3 above)
Comment (sezero) [reply src/loaders/asylum_load.c:150]: ok.

File: src/hmn_extras.c (hunk 2 above)
Comment (sezero) [reply src/hmn_extras.c:134]: OK then, let's get this in, hopefully nothing breaks.

**Review by sezero — APPROVED** [review]
OK by me - waiting for you to find / fix that possible `read_event_ft2` issue.

**Comment by AliceLR** [comment]
> OK by me - waiting for you to find / fix that possible `read_event_ft2` issue.

I did find out it's a segmentation fault in the `FX_ENVPOS` handler related to `read_event_ft2`, but not what causes it. Will look into it now.

**Comment by AliceLR** [comment]
`libxmp_get_instrument` (called by the `FX_ENVPOS` handler) was indexing `mod->xxi` by an `xc->ins` value of -1, which means no instrument or an invalid instrument.

**Comment by AliceLR** [comment]
Ugh, it looks like `libxmp_get_sample` might have the same bug, will push another patch once I've verified it.

**Pull request merged**